conwayste/src/ui/button.rs

```diff
@@ -0,0 +1,187 @@
+/*  Copyright 2019 the Conwayste Developers.
+ *
+ *  This file is part of conwayste.
+ *
+ *  conwayste is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  conwayste is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with conwayste.  If not, see
+ *  <http://www.gnu.org/licenses/>. */
+
+use std::rc::Rc;
+
+use chromatica::css;
+
+use ggez::graphics::{self, Rect, Font, Color, DrawMode, DrawParam};
+use ggez::nalgebra::{Point2, Vector2};
+use ggez::{Context, GameResult};
+
+use super::{
+    label::Label,
+    widget::Widget,
+    helpe::{within_widget, color_with_alpha, center},
+    UIAction, WidgetID
+};
+
+pub struct Button {
+    pub id:     WidgetID,
+    pub label: Label,
+    pub button_color: Color,
+    pub draw_mode: DrawMode,
+    pub dimensions: Rect,
+    pub hover: bool,
+    pub borderless: bool,
+    pub action: UIAction
+}
+
+const BUTTON_LABEL_PADDING_W: f32 = 16.0;   // in pixels
+const BUTTON_LABEL_PADDING_H: f32 = 16.0;   // in pixels
+
+/// A named widget that can be clicked to result in an occuring action.
+impl Button {
+
+    /// Creates a Button widget.
+    ///
+    /// # Arguments
+    /// * `ctx` - GGEZ context
+    /// * `font` - font to be used when drawing the text
+    /// * `button_text` - Text to be displayed
+    /// * `widget_id` - Unique widget identifier
+    /// * `action` - Unique action identifer
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use ggez::graphics::Font;
+    /// use ui::Button;
+    ///
+    /// fn new(ctx: &mut Context) -> GameResult<MainState> {
```

Comment:
Is this a valid example? It returns `GameResult<MainState>`.

(this goes for the other widget examples too)

Reply:
Not intended to be a full example, but merely code to show you how to create and draw a button. It returns `GameResult<MainState>` because that's what `draw` returns, and I'm keeping inline with the ggez API.